EpikCord/__init__.py

```diff
@@ -4186,4 +4188,90 @@ async def default_failure(self, interaction):
 class CommandUtils:
     
     def check(self, callback):
-        return Check(callback)
\ No newline at end of file
+        return Check(callback)
+
+class BaseConnectable:
+    def __init__(self,client:Client):
+        self.client = client
+        self.ws = None
+        self.IDENTIFY = 0
+        self.SELECT_PROTOCOL = 1
+        self.READY = 2
+        self.HEARTBEAT = 3
+        self.HELLO = 8
+        self.connected = False
+        self.server_set = False
+        self.state_set = False
+        self.sequence = None
+        self.endpoint:str = None
+        #have to add events manually
+        self.client.events["voice_server_update"] = self.voice_server_update
+        self.client.events["voice_state_update"] = self.voice_state_update
+    async def voice_state_update(self,data):
+        if self.connected and data["channel_id"]== self.channel_id:
+            self.session_id = data["session_id"]
+            self.state_set = True
+
+
+    async def voice_server_update(self, data):
+        if self.connected and data["guild_id"] == self.guild_id:
+            self.token = data["token"]
+            self.guild_id = data["guild_id"] #ain't gonna make a difference
+            self.endpoint = data["endpoint"]
+            self.server_set = True
+    async def connect(self, guild_id:str, channel_id:str, self_mute:bool = False, self_deaf:bool = False):
+        
+        await self.client.send_json({
+            "op":4,
+            "d":{
+                "guild_id":guild_id,
+                "channel_id":channel_id,
+                "self_mute":self_mute,
+                "self_deaf": self_deaf
+            }
+
+        })
+        self.connected = True
+        self.guild_id = guild_id
+        self.channel_id = channel_id
+        if self.endpoint.startswith("wss://"):
+            self.endpoint = self.endpoint[6:]
+        self.ws = await self.client.http.ws_connect(f"wss://{self.endpoint}")
+        await self.send_json({
+            "op": self.IDENTIFY,
+            "d": {
+                "server_id": self.guild_id,
+                "user_id": self.client.user.id,
+                "session_id": self.session_id,
+                "token": self.token
+            }
+        })
+
+    async def udp_start(self):
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+        self.socket.setblocking(False)
+        self.socket.sendto(b'', (self.ip, self.port))
+        self.final_mode = random.choice(self.modes)
```

Comment:
What are the modes and why are them chosen randomly?

Reply:
> What are the modes

https://discord.com/developers/docs/topics/voice-connections#establishing-a-voice-udp-connection-encryption-modes

Reply:
well they are encryption modes which encrypt audio
They are chosen randomly bc security or whatev

Reply:
Secret key is not implemented?

Reply:
> [Example Session Description Payload](https://discord.com/developers/docs/topics/voice-connections#establishing-a-voice-udp-connection-example-session-description-payload)

```js
{
    "op": 4,
    "d": {
        "mode": "xsalsa20_poly1305_lite",
        "secret_key": [ ...251, 100, 11...]
    }
}
```

Reply:
Nah it aint implemented, he has coded something like basic ones.

Reply:
later after udp